plugins/tplink/tl_sg108e

```diff
@@ -0,0 +1,210 @@
+            }
+            else {
+                print "graph_vlabel packets\n";
+                foreach my $gb (qw/good bad/) {
+                    foreach my $direction (qw/tx rx/) {
+                        print "p${port}${direction}${gb}.label Port $port $direction ($gb)\n";
+                        print "p${port}${direction}${gb}.type COUNTER\n";
+                    }
+                }
+            }
+        }
+    }
+    exit 0;
+}
+
+# extract data from the switch - uses web scraping (tested with 1.0.2 Build 20160526 Rel.34615)
+
+# print STDERR "Connecting to $switch with user $username";
+my $mech = WWW::Mechanize->new(
+    autocheck             => 0,
+    requests_redirectable => [ 'GET', 'HEAD', 'POST' ],
+    user_agent => 'Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0'
+);
+my $result = $mech->post(
+    "http://$switch/logon.cgi",
+    [ username => $username, password => $password, logon => 'Login' ],
+    "Referer" => "http://$switch/"
+);
+my $response = $mech->response();
+
+# navigate to the page with the network stats
+$result   = $mech->get("http://$switch/PortStatisticsRpm.htm");
+$response = $mech->response();
+
+# print STDERR $response->code()."\n";
+
+# get the data
+my $data = $mech->content( raw => 1 );
+
+# print STDERR "$data\n";
+
+# The page stores the data in a table, but internally it is stored in 3 javascript arrays:
+# state:[1,1,1,1,1,1,1,1,0,0],
+# link_status:[0,5,0,0,5,5,5,6,0,0],
+# pkts:[0,0,0,0,14141090,0,10461386,0,14226,0,12252,0,0,0,0,0,2872063,0,1402200,0,59764503,0,34619246,0,4913873,0,4393574,0,44170456,0,68499653,0,0,0]
+
+# state: 1 - Enabled, 0 - Disabled (administratively)
+# link_status: 0 - down, 5 - 100Mbps full, 6 - 1Gbps
+# pkts: every group of 4 values represent txGoodPkt, txBadPkt, rxGoodPkt, rxBadPkt
+
+# parse good/bad packets
+if ( $data =~ /pkts:\[([0-9,]+)\]/ ) {
+
+    my $packetString = $1;
+    my @packets = split( /,/, $packetString );
+
+    for ( 1 .. $numberOfPorts ) {
+        my $currentPort = $_;
+        my $txGoodPkt   = $packets[ ( $currentPort - 1 ) * 4 ];
+        my $txBadPkt    = $packets[ ( $currentPort - 1 ) * 4 + 1 ];
+        my $rxGoodPkt   = $packets[ ( $currentPort - 1 ) * 4 + 2 ];
+        my $rxBadPkt    = $packets[ ( $currentPort - 1 ) * 4 + 3 ];
+        print "multigraph Packets_if_$currentPort\n";
+
+        print "p${currentPort}txgood.value $txGoodPkt\n";
+        print "p${currentPort}rxgood.value $rxGoodPkt\n";
+        print "p${currentPort}txbad.value $txBadPkt\n";
+        print "p${currentPort}rxbad.value $rxBadPkt\n";
+    }
+}
+
+# parse link speed
+if ( $data =~ /link_status:\[([0-9,]+)\]/ ) {
+
+    my $linkString = $1;
+    my @links = split( /,/, $linkString );
+    for ( 1 .. $numberOfPorts ) {
+        my $currentPort = $_;
+        my $link        = $links[ $currentPort - 1 ];
+        print "multigraph Speed_if_$currentPort\n";
+
+        print "p${currentPort}.value $link\n";
```

Comment:
Please transform the device-specific metric (0/5/6) into a human-readable metric here.
How about `0` for `not connected`, `100` for `5` and `1000` for `6`?
All other values could be represented by `U`.

Reply:
Sure, that's sensible. However, the values are not only speed, but speed/duplex (e.g. 5 is 100 Full Duplex, 6 is 1000 Full Duplex. Probably 4 is 100 Half Duplex.). I will need to map them out, but I'm not sure how to display them.

Reply:
Since you are visualizing distinct states without a real numerical representation, you could also just define a number of states and set the value to zero or one, if the state is currently active:
```
# config
speed_100_half_duplex.label 100 MBit/s Half Duplex
speed_100_full_duplex.label 100 MBit/s Full Duplex
speed_1000_full_duplex.label 1 GBit/s Full Duplex

# values
speed_100_half_duplex.value 0
speed_100_full_duplex.value 1
speed_1000_full_duplex.value 0
```

The resulting graph would be similar to this one: [channel used by a wireless device](https://monitor.opennet-initiative.de/munin/ap/AP2.76/wireless_channel_active_phy0.html)